add equiangular EXR IBL (Image based lighting) example

richardmonette opened the pull request:
Fixes https://github.com/mrdoob/three.js/issues/13574

<img width="816" alt="screen shot 2018-03-26 at 4 17 03 pm" src="https://user-images.githubusercontent.com/1041278/37930947-2d336666-3112-11e8-8873-9a7c38c192a6.png">

#### Summary

This PR adds an example of an IBL (Image based lighting) pipeline based around an equiangular formatted EXR. 

Since we already have a nice `PMREMGenerator`, I have taken the approach of adding a `EquiangularToCubeGenerator` which uses the GPU to convert an equiangular format texture into a cube map. The highlight here is the equiangular EXR workflow, but I have also included in the example using a PNG equiangular map, just to demonstrate the `EquiangularToCubeGenerator` conversion is also possible in a low dynamic range context.

#### Performance

Test

```
console.time('EquiangularToCubeGenerator');

var cubemapGenerator = new THREE.EquiangularToCubeGenerator( texture, 512 );
var cubeMapTexture = cubemapGenerator.generate( renderer );

console.timeEnd('EquiangularToCubeGenerator');
```

Result

```
EquiangularToCubeGenerator: 28.334716796875ms
```

@mrdoob @WestLangley

richardmonette committed:
add equiangular exr IBL example

looeee commented:
[Live version](https://rawgit.com/richardmonette/three.js/dde7310aaeceb2485c578ba4c1b4fc7788973ba9/examples/webgl_materials_equiangular_ibl.html). Looks amazing btw 😍 

Is there a reason why PNG seems so overexposed compared to EXR?

richardmonette commented:
@looeee Without getting super into the details, perhaps suffice to say the PNG version is using LDR (low dynamic range) so its never going to look correct, you can mess with the exposure a bit to adjust, but the primary goal here is to establish the correct pipeline using HDR (high dynamic range) light probes as the basis of the IBL (image base lighting) pipeline.

WestLangley commented on examples/webgl_materials_equiangular_ibl.html:
```diff
@@ -0,0 +1,236 @@
+
+				standardMaterial = new THREE.MeshStandardMaterial( {
+					map: null,
+					color: 0x000000,
```
Why black? I suggest white -- and reduce the default exposure, accordingly

WestLangley commented on examples/webgl_materials_equiangular_ibl.html:
```diff
@@ -0,0 +1,236 @@
+				scene.add( torusMesh1 );
+				objects.push( torusMesh1 );
+
+				floorMaterial = new THREE.MeshStandardMaterial( {
```
I think `MeshBasicMaterial` is more appropriate, here, as the purpose is only to display the mips.

WestLangley commented:
Sweet!

For a future PR, if you have any ideas on how to reduce the memory stats, that would be great.

    memory:
        geometries: 318
        textures: 18

richardmonette committed:
adjust materials

richardmonette commented:
@WestLangley Thank you for your comments! I've made the changes you recommended and I think it even slightly improves the visual quality. Good call with the MeshBasicMaterial, it unlit rendering is exactly what we needed for the debug texture.

WestLangley commented:
@looeee I took the liberty of editing your temporary live link as it was not accommodating the additional commits. :)

mrdoob commented:
@WestLangley Thanks! The new link looks much better indeed!

Mugen87 commented on examples/js/loaders/EquiangularToCubeGenerator.js:
```diff
@@ -0,0 +1,111 @@
+	];
+
+	this.camera = new THREE.PerspectiveCamera( 90, 1, 0.1, 10 );
+	this.planeMesh = new THREE.Mesh( new THREE.BoxGeometry( 1, 1, 1 ), this.getShader() );
```
I would choose `BoxBufferGeometry` here.

Mugen87 commented on examples/webgl_materials_equiangular_ibl.html:
```diff
@@ -0,0 +1,231 @@
+					envMapIntensity: 1.0
+				} );
+
+				var geometry = new THREE.TorusKnotGeometry( 18, 8, 150, 20 );
```
And here: `TorusKnotBufferGeometry`

Mugen87 commented:
I have a black screen and the following warnings in the console. 
```
[.Offscreen-For-WebGL-0x55fb20407c10]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glClear: framebuffer incomplete (check)
[.Offscreen-For-WebGL-0x55fb20407c10]GL ERROR :GL_INVALID_FRAMEBUFFER_OPERATION : glDrawArrays: framebuffer incomplete (check)
```
Tested with Chromium 64.0.3282.167 on Ubuntu 16.04. Other HDR examples (like https://threejs.org/examples/webgl_materials_envmaps_hdr.html) seem to work:

richardmonette committed:
switch to bufferGeo, use RGBAF (better supported)

richardmonette commented:
@Mugen87 Thank you for your review. I have made the following changes, based upon your suggestions:

- Changed both of the suggested geometries to buffers
- I have fixed the issue with FF 59.0.1 (referring to https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_color_buffer_float I observed that the render target appears to _only_ support RGBA32F_EXT, and RGB32F_EXT is deprecated, so I have adjusted accordingly)

Would it be possible for you to please confirm that's working for you too now? Thanks!

Mugen87 commented:
Great :+1: . The example now works in Chrome and FF :tada:

WestLangley commented on examples/js/loaders/EquiangularToCubeGenerator.js:
```diff
@@ -0,0 +1,111 @@
+    			\n\
+					gl_FragColor = vec4( color, 1.0 );\n\
+				}",
+			blending: THREE.CustomBlending,
```
Can you please explain the reason for the custom blending?

richardmonette replied:
Hey @WestLangley , I'm not specifically familiar with the `THREE.ShaderMaterial` so I just took that from an example. If its an option, this could probably just be left as the defaults.

WestLangley commented on examples/js/loaders/EquiangularToCubeGenerator.js:
```diff
@@ -0,0 +1,111 @@
+
+	this.camera = new THREE.PerspectiveCamera( 90, 1, 0.1, 10 );
+	this.planeMesh = new THREE.Mesh( new THREE.BoxBufferGeometry( 1, 1, 1 ), this.getShader() );
+	this.planeMesh.material.side = THREE.DoubleSide;
```
Why is it `planeMesh` and why is it double-sided?

richardmonette replied:
`planeMesh` should be renamed to `boxMesh`, it's double sided because the camera is inside the box, i.e. it needs to render and not be backface culled.

WestLangley replied:
side: THREE.BackSide;

As you know, users copy code, so let's make it clear. :)

richardmonette replied:
👍(btw, appreciate you taking the time to work through the code with me!)

WestLangley commented on examples/js/loaders/EquiangularToCubeGenerator.js:
```diff
@@ -0,0 +1,111 @@
+	this.planeMesh.material.side = THREE.DoubleSide;
+	this.scene = new THREE.Scene();
+	this.scene.add( this.planeMesh );
+	this.scene.add( this.camera );
```
I don't think adding the camera to the scene is required.

WestLangley commented on examples/js/loaders/EquiangularToCubeGenerator.js:
```diff
@@ -0,0 +1,111 @@
+
+			this.camera.position.set(0, 0, 0);
+			this.camera.up.set(v.u[0], v.u[1], v.u[2]);
+			this.camera.lookAt(new THREE.Vector3(v.t[0], v.t[1], v.t[2]));
```
this.camera.lookAt( v.t[ 0 ], v.t[ 1 ], v.t[ 2 ] );

richardmonette committed:
rename boxMesh, remove custom blending

richardmonette committed:
dont need camera in scene, can avoid a new

WestLangley commented:
How are you handing the encoding? Are you assuming `.exr` is in linear space? `.png` in `sRGB`?

richardmonette commented:
EXR is in linear, yes.

richardmonette committed:
be specific about using BackSide

WestLangley commented:
>Is there a reason why PNG seems so overexposed compared to EXR?

Likely the encoding.

Can you explicitly set the proper encodings in your example? `THREE.sRGBEncoding` for `.png`, `THREE.LinearEncoding` for `.exr`.

richardmonette commented:
> Unrelated to this PR, but I think there is something wrong with the pmrem code. By changing the roughness of the material we can see that the levels are not aligned (the lights in the reflection changes position).

https://github.com/mrdoob/three.js/blob/dev/examples/js/pmrem/PMREMGenerator.js#L103

That's something I am looking at in my next PR... 😄

mrdoob commented:
> That's something I am looking at in my next PR... 😄

Sweet!

I think this PR only needs @WestLangley last comment addressed and it's ready to go.

richardmonette committed:
add explicit encoding

richardmonette commented:
@WestLangley I've explicitly set the proper encodings as you suggested and things have come even further into line 👍

<img width="500" alt="hdr vs ldr" src="https://user-images.githubusercontent.com/1041278/37997762-fddc10f6-31e9-11e8-8748-c5036eca2d3f.png">

You'll notice when adjusting exposure (especially where the EXR has more detail in the shadows) that the EXR / high dynamic range version works better, but there isn't the strange visual difference anymore 👌

WestLangley commented:
Yes, the rendering is looking more reasonable now.   :)

What I don't yet understand is why `generateMipmaps` is `false` for `.exr` and `true` for `.png` -- and what it _should_ be. That can wait for another PR, though.

/ping @bhouston

bhouston commented:
> What I don't yet understand is why generateMipmaps is false for .exr and true for .png -- and what it should be. That can wait for another PR, though.

Actually it isn't because of EXR or PNG but rather RGBE can not be linearly interpolated as it isn't a monotonic encoding, where as RGBM is a relatively monotonic encoding.

bhouston commented:
> Unrelated to this PR, but I think there is something wrong with the pmrem code. By changing the roughness of the material we can see that the levels are not aligned (the lights in the reflection changes position).

This is caused by the sample pattern not being fully symmetric.

bhouston commented:
If the PNG in this example is truly LDR and there is no encoding (e.g. RGBE, RGBM, RGBD, etc.) then it can be linear interp as well like the EXR HDR.

richardmonette commented:
> This is caused by the sample pattern not being fully symmetric.

I've got some ideas about using a low discrepancy sampler, such as a Hammersley, to improve quality as well.

bhouston commented:
> I've got some ideas about using a low discrepancy sampler, such as a Hammersley, to improve quality as well, for a follow up PR.

Okay.  I'm interested.  I think that the issue is the current reliance on a GPU-based random number generator which is not uniform, nor is it consistent across different GPU families.  I once had hope that there could be a reliable, high quality GPU-based "random" number generator -- test cases here https://threejs.org/examples/#webgl_postprocessing_procedural -- but I have since given up hope.   I think a Hammersley distribution or a 2D Poisson disk would work great here. 

You also hint that you can improve the theoretical foundation of this convolver, that would be really amazing.  The problem with this one is that it uses the result of the previous level to create the next to get speed and reduce the number of samples required, thus it is a little tricky.

bhouston commented:
I would rename this example to webgl_materials_envmaps_exr.html so that it is beside the existing weblg_materials_envmaps_hdr.html and weblg_materials_envmaps.html as they are all related.

richardmonette committed:
rename examples/webgl_materials_equiangular_ibl.html -> examples/webgl_materials_envmaps_exr.html

mrdoob commented:
Thanks!

bhouston commented:
This was a great PR and work @richardmonette, very impressive.
